File: python/easy/303.range_sum_query_inmutable/range_sum_query_array_inmutable.py
from typing import List

class NumArray:
    nums = []

    def __init__(self, nums: List[int]):
        if len(nums) == 0:
            return

        prefixSum = []
        prefixSum.append(nums[0])
        for i in range(1, len(nums)):
            prefixSum.append(prefixSum[i-1] + nums[i])

        self.nums = prefixSum
        

    def sumRange(self, left: int, right: int) -> int:
        if left == 0:
            return self.nums[right]

        return self.nums[right] - self.nums[left-1]
        


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(left,right)

File: c++/easy/303.range_sum_query_inmutable/range_sum_query_array_inmutable.cpp
class NumArray {
private:
    vector<int> nums;
public:
    NumArray(vector<int>& nums) {
        if(nums.size() == 0) {
            return;
        }

        vector<int> prefixSum(nums.size());
        prefixSum[0] = nums[0];
        for(int i = 1; i < nums.size(); i++) {
            prefixSum[i] = prefixSum[i-1] + nums[i];
        }

        this->nums = prefixSum;
    }
    
    int sumRange(int left, int right) {
        if(left == 0) {
            return this->nums[right];
        }

        return this->nums[right] - this->nums[left-1];
    }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * int param_1 = obj->sumRange(left,right);
 */